File: scripts/interop/clients/typescript/interop.mjs
// Interop cell: TypeScript SDK (@hivehub/thunder) against a Thunder-based server.
//
// Driven by scripts/interop/run-matrix.py. Prints one
// `STEP <name> PASS|FAIL <detail>` line per step, exits non-zero on any failure.
//
// Runs against the SDK's built `dist/`, i.e. what a consumer installs, rather
// than against the TypeScript sources.

import { SynapClient } from '../../../../sdks/typescript/dist/index.js';

// Not valid UTF-8, so a transport that quietly round-trips through a string
// cannot pass the binary step.
const BINARY = Buffer.from([0xde, 0xad, 0xbe, 0xef]);
const TOPIC = 'interop.typescript';

let failures = 0;

function report(step, ok, detail) {
  console.log(`STEP ${step} ${ok ? 'PASS' : 'FAIL'} ${detail}`);
  if (!ok) failures += 1;
}

const [host, port, user, pass] = process.argv.slice(2);

const client = new SynapClient({
  url: `synap://${host}:${port}`,
  timeout: 15000,
  auth: { type: 'basic', username: user, password: pass },
});

const rpc = client.synapRpcTransport();
if (!rpc) {
  report('auth', false, 'synap:// URL did not select the SynapRPC transport');
  process.exit(1);
}

// 1. Authenticate. Credentials ride the handshake on the first call; the
//    pre-Thunder transport never sent AUTH, so a require_auth server was
//    unreachable.
//
//    EXISTS rather than PING: the server answers PING before authentication,
//    so a PING probe passes just as happily on a connection that never
//    authenticated -- which is exactly the bug this column exists to catch.
try {
  const probe = await rpc.execute('EXISTS', ['interop:ts:probe']);
  report('auth', true, `EXISTS -> ${JSON.stringify(probe)}`);
} catch (err) {
  report('auth', false, `${err.constructor.name}: ${err.message}`);
  process.exit(1);
}

// 2. SET/GET a binary value — canonical MessagePack bin, byte-exact back.
try {
  await rpc.execute('SET', ['interop:ts:bin', BINARY]);
  const got = await rpc.execute('GET', ['interop:ts:bin']);
  const bytes = Buffer.isBuffer(got) ? got : Buffer.from(got, 'binary');
  report('kv_binary', bytes.equals(BINARY), `${BINARY.toString('hex')} -> ${bytes.toString('hex')}`);
} catch (err) {
  report('kv_binary', false, `${err.constructor.name}: ${err.message}`);
}

// 3. SUBSCRIBE then PUBLISH — the push frame must arrive on the hook.
let subscription;
try {
  const received = [];
  const arrived = new Promise((resolve) => {
    subscription = rpc.subscribePush([TOPIC], (msg) => {
      received.push(msg);
      resolve();
    });
  });
  subscription = await subscription;

  await rpc.execute('PUBLISH', [TOPIC, 'interop-payload']);
  await Promise.race([arrived, new Promise((r) => setTimeout(r, 10000))]);

  const ok = received.length > 0 && received[0].topic === TOPIC;
  report('pubsub', ok, `received=${JSON.stringify(received.slice(0, 1))}`);
} catch (err) {
  report('pubsub', false, `${err.constructor.name}: ${err.message}`);
} finally {
  subscription?.cancel?.();
}

// 4. Error round-trip — an unknown command must reject, and must not poison
//    the multiplexed connection.
try {
  const result = await rpc.execute('NOSUCHCOMMAND', []);
  report('error', false, `expected a rejection, got ${JSON.stringify(result)}`);
} catch (err) {
  const alive = (await rpc.execute('PING', [])) === 'PONG';
  report('error', alive, `rejected with ${err.constructor.name}; connection alive=${alive}`);
}

rpc.close();
client.close?.();
process.exit(failures ? 1 : 0);

File: scripts/interop/clients/python/interop.py
#!/usr/bin/env python3
"""Interop cell: Python SDK (hivellm-thunder) against a Thunder-based server.

Drives ``SynapRpcTransport`` directly rather than the sugar layer -- the matrix
is about the wire, and the transport is where the wire lives.
"""

from __future__ import annotations

import asyncio
import sys
import threading

from synap_sdk.exceptions import SynapException
from synap_sdk.transport_rpc import SynapRpcTransport
from thunder_rpc import Credentials

# A value that is emphatically not valid UTF-8, so a transport that quietly
# round-trips through a string cannot pass this cell.
BINARY = bytes([0xDE, 0xAD, 0xBE, 0xEF])
TOPIC = "interop.python"


def report(step: str, ok: bool, detail: str) -> None:
    print(f"STEP {step} {'PASS' if ok else 'FAIL'} {detail}", flush=True)


async def main() -> int:
    host, port, user, password = sys.argv[1], int(sys.argv[2]), sys.argv[3], sys.argv[4]
    failures = 0

    transport = SynapRpcTransport(
        host, port, timeout=15.0, credentials=Credentials.user_pass(user, password)
    )

    # 1. authenticate -- the handshake happens on the first call.
    #
    # EXISTS rather than PING: the server answers PING before authentication,
    # so a PING probe passes just as happily on a connection that never
    # authenticated -- exactly the bug this column exists to catch.
    try:
        probe = await transport.execute("EXISTS", ["interop:python:probe"])
        report("auth", True, f"EXISTS -> {probe!r}")
    except SynapException as exc:
        report("auth", False, f"{type(exc).__name__}: {exc}")
        await transport.close()
        return 1

    # 2. SET/GET a binary value -- canonical MessagePack bin, byte-exact back.
    try:
        await transport.execute("SET", ["interop:python:bin", BINARY])
        got = await transport.execute("GET", ["interop:python:bin"])
        got_bytes = got.encode("latin-1") if isinstance(got, str) else got
        ok = got_bytes == BINARY
        report("kv_binary", ok, f"{BINARY.hex()} -> {bytes(got_bytes or b'').hex()}")
        failures += 0 if ok else 1
    except SynapException as exc:
        report("kv_binary", False, f"{type(exc).__name__}: {exc}")
        failures += 1

    # 3. SUBSCRIBE then PUBLISH -- the push frame must arrive on the hook.
    received: list[dict] = []
    arrived = threading.Event()
    cancel = None
    try:
        def on_message(msg: dict) -> None:
            received.append(msg)
            arrived.set()

        _subscriber_id, cancel = await transport.subscribe_push([TOPIC], on_message)
        await transport.execute("PUBLISH", [TOPIC, "interop-payload"])

        for _ in range(50):
            if arrived.is_set():
                break
            await asyncio.sleep(0.1)

        ok = bool(received) and received[0]["topic"] == TOPIC
        report("pubsub", ok, f"received={received[:1]}")
        failures += 0 if ok else 1
    except SynapException as exc:
        report("pubsub", False, f"{type(exc).__name__}: {exc}")
        failures += 1
    finally:
        if cancel is not None:
            cancel()

    # 4. Error round-trip -- an unknown command must surface as an exception,
    #    not as a null result, and must not poison the connection.
    try:
        result = await transport.execute("NOSUCHCOMMAND", [])
        report("error", False, f"expected an error, got {result!r}")
        failures += 1
    except SynapException as exc:
        still_alive = await transport.execute("PING", [])
        ok = still_alive == "PONG"
        report("error", ok, f"raised {type(exc).__name__}; connection alive={ok}")
        failures += 0 if ok else 1

    await transport.close()
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
